### ISSUE TITLE
Should we hard wrap commit messages?

### ISSUE THREAD

**Issue by joshaber** [desktop/desktop#62]
Enter a commit message and commit.

**Comment by jacobbednarz** [comment]
+1 on making the client hard wrap the commits to be consistent with GitHub & good practices from Git itself.

**Comment by redhotvengeance** [comment]
I found this issue after being dismayed that the new GH Desktop app _doesn't_ wrap, which I was expecting it to do since that's what the previous version did.

At the core of GitHub is Git, and I think it makes sense for GitHub to encourage people to follow Git's best practices. It is widely accepted that Git commit messages should:

- Have a title no longer than 50 chars
- Have an optional body wrapped at 72 chars

Therefore, I think it makes sense for GitHub to enforce these practices at well. Or, at the very least, have their tools optionally enable/disable them. While I primarily interact with Git via the CLI, I enjoy using GitHub Desktop to check diffs, do line-based commits, etc. If GitHub Desktop doesn't properly wrap, I'll have to bounce back to the CLI to compose all commits with a body.

At the end of the day, Git commits should:

- Follow best community practices
- Look good in the terminal (where Git lives)

**Comment by danny-andrews** [comment]
Any update on this? The lack of line length indicators on the desktop app makes it almost unusable.

**Comment by danny-andrews** [comment]
Having a twitter-like counter would be a good solution, I think.

**Comment by agatac** [comment]
> The lack of line length indicators on the desktop app makes it arduous to use as I have to copy and paste commit messages into an editor to check line lengths.

I came here to suggest just that. Length indicator would be useful. So much easier to follow Git's best practices.

**Comment by shiftkey** [comment]
We probably need to think about how small the current commit form is before we can show any fancy UI here. Related issue: #1646

**Comment by dhoulb** [comment]
I'd prefer no hard wrapping, i.e. that GitHub Desktop had a forward-looking approach to Git where wrapping is handled when the message is displayed, not by stealthily modding user input.

I also can't see how it'd be implemented without _disabling_ the current softwrap on that textarea so lines don't wrap, which would be unexpected behaviour for most newbie users. So it probably needs to be enabled by an advanced setting checkbox.

Feels like a more complete solution would be to include this in some potential future functionality that also implements commit templates, so maintainers of projects could specify how they want commits formatted, wrapped, and structured.

(However: I would love a soft character-limit reminder on the title! Shorter titles significantly improve the readability of commit lists. Just painting the excess characters in red or warning orange etc would be enough!)

**Comment by fatcerberus** [comment]
> I'd prefer no hard wrapping, i.e. that GitHub Desktop had a forward-looking approach to Git where wrapping is handled when the message is displayed, not by stealthily modding user input.

While I agree with the basic sentiment, I don't think it's common for command-line apps to do any kind of fancy wrapping like this, in fact half the reason to drop to the command line in the first place is to see things exactly as they are (terminal window size constraints aside) without any of the smoke and mirrors of GUIs.  So it would be weird for a command-line app like Git to auto-format things for display, I think.  Therefore, while this approach would improve the experience for users of GUIs (and the GH website), it would greatly degrade the experience for all users accessing your repository from the command line (i.e. the default Git workflow).

I don't think the client needs any kind of fancy line length indicators in the text area either, I would be just as happy with a simple checkbox to enable automatic hard wrapping, perhaps with a user-customizable line length.

**Comment by itskingori** [comment]
My 2 cents, line length indicator would be enough. Possibly faint grey to give the impression that it's a guide and not necessarily a hard rule. On hard-wrapping, I find it idealistic ... in the sense that not everything can be hard-wrapped e.g. a long URL.

**Comment by fatcerberus** [comment]
The main reason I like that the old client did wrapping after the fact is that when composing the message I can make edits to the text freely, without having to shuffle around the line breaks.  Then once I commit they’re all in the right place when viewing the commit on the command line.  AND it allows the textbox to use a variable width font to ease reading.  That’s a good GUI experience and is the main reason I preferred the old client for day-to-day work over the command line.

Good point for things like URLs, but I’d also note that any word-wrapping algorithm worth its salt *won’t* break a line with no whitespace.

**Comment by HANHTHONG7985** [comment]
NHUNG THU NAY CO THE LAM GI? CO RA TIEN K?


2018-03-17 21:51 GMT+07:00 Bruce Pascoe <notifications@github.com>:

**Comment by senterright** [comment]
I think @fatcerberus hits all the right notes. the main points for me are

- git should be the "source of truth", github desktop should make things nicer
- vim literally does this for you (if you use that as your commit message editor)
- checkboxes seem like they'd work for both validation/length warnings as well as just automatically formatting (as @fatcerberus points out, a good algo should probably only break on whitespace, again, see vim)

for now i'm ok with going into the command line and doing the 72 char limit formatting myself. it's actually not that big a deal for me personally. right now it's still slightly better than doing it with vim currently (which itself has its issues like cut/copy/paste or adding to a line that's already past the 72 char limit) because i can just do all the shuffling after i have finished the message instead of doing it while i'm trying to write the message too.

github desktop has come a long way! grats to everyone for making it super dope and i'm sure we're all looking forward to future enhancements.

**Comment by sandstrom** [comment]
This seems to exist in the Github Atom package:
https://github.com/atom/github/pull/1223

Would be great if it could be added to desktop too.

**Comment by sandstrom** [comment]
@shiftkey Any update on this? Right now using Github Desktop causes "pollution" in the git log, with commit messages without line breaks.

Since the Git workflow is pretty core to Github, and since it's a very established practice to wrap git commit messages, wouldn't it make sense to prioritize this issue from Githubs end?

*(I understand that it may not all be up to you, but you're working there so I'll still ask you 😄)*

**Comment by JKRhb** [comment]
I opened PR #12421 with a suggestion for word-wrapped commit descriptions. This approach might be a bit too straightforward, though, as there probably should be a way to enable or disable this feature. In any case, some feedback/discussion on this would be awesome :) The approach linked by @sandstrom looks very promising to me.

**Comment by DaveCousineau** [comment]
I've been thinking of this for a while. Why are we encouraged to do this? Shouldn't it be _bad_ practice to "hard wrap" text, especially in modern times? Wrapping text is a presentation issue. I would think that the raw text itself should not be wrapped, and that whatever it is that happens to display it should decide how to word wrap it, if and as necessary. Why should I presuppose the width of the display that the text will be displayed on and bake that assumption into the text itself?

**Comment by JKRhb** [comment]
In the meantime, I also think that users should not be forced to have their commit descriptions hard-wrapped. However, a warning like the one you currently get for the title would be nice to have. Maybe the text could also turn red, as it does when using Git on the command line if a certain line length has been reached. This line length could be configurable, with the option to completely disable the feature.

**Comment by azeemba** [comment]
I understand the hesitance with forcing the wrapping on the user but I would appreciate a keyboard shortcut or a button to wrap the current message.

That way folks that don't care don't have to change but people that do work on repos that require pre-wrapped-messages still get the help.

**Comment by ajm19826** [comment]
anyone still here in 2025?

**Comment by niik** [comment]
Hi folks, we discussed this issue in our last issue review meeting and we've decided against working on this feature at this time and as such we're going to close this issue.

We recognize that fine grained control over the formatting of commit messages are important to some of you and we're not ruling out a more advanced commit message editor in the future (loose ideas we talked about was a dedicated commit message writing dialog where we could offer more of an editor-like experience with guidelines and such) but we don't believe automatically hard wrapping is the solution here.

With the recent work to support Git hooks in GitHub Desktop we hope that could be a solution for at least some of you. Either a validating pre-commit hook that rejects commit not matching your expected format or a hook that automatically performs the hard-wrapping preferences your project enforces.
